### PR TITLE
feat: Ability to limit request size and connection count 

### PR DESCRIPTION
**Pull request by RudraniW**
### **Description**
 
This PR allows a requestSizeLimit (maximum bytes for a request, including the header size) and connectionLimit (total number of concurrent connections) to be configured when registering a server.

Example usage:
`let serverConfig = HTTPServerConfiguration(requestSizeLimit: 1024, connectionLimit: 1024)`

If you do not specify either of the parameters, or do not specify a`HTTPServerConfiguration` at all, then default values are used. The defaults are defined in `HTTPServerConfiguration` in Kitura-NIO here: [https: //github.com/RudraniW/Kitura-NIO/blob/requestSize/Sources/KituraNet/HTTP/HTTPServerConfiguration.swift](url)

### **Motivation and Context**
See: #206 and https://github.com/IBM-Swift/Kitura/issues/1384

### CODE REVIEW

File: Sources/KituraNet/HTTP/HTTPServer.swift [hunk 1]
```diff
@@ -660,3 +673,23 @@ enum KituraWebSocketUpgradeError: Error {
     // Unknown upgrade error
     case unknownUpgradeError
 }
+
+class Atomic<A> {
```
Comment (weissi) [inline Sources/KituraNet/HTTP/HTTPServer.swift]: you could use [NIO's Atomic<T>](https://apple.github.io/swift-nio/docs/current/NIOConcurrencyHelpers/Classes/Atomic.html) which actually is atomic and doesn't require a lock. Just add `NIOConcurrencyHelpers` to `Package.swift` and `import NIOConcurrencyHelpers`

**Comment by RudraniW** [comment]
I am closing this PR as I have implemented the other approach to limit request size and connection count in #222 .

File: Sources/KituraNet/HTTP/HTTPServerConfiguration.swift [hunk 2]
```diff
@@ -0,0 +1,43 @@
...
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+import Foundation
+
+public struct HTTPServerConfiguration {
```
Comment (djones6) [inline Sources/KituraNet/HTTP/HTTPServerConfiguration.swift]: The naming and facilities on this struct need to be consistent between Kitura-net and Kitura-NIO.  In https://github.com/IBM-Swift/Kitura-net/pull/307 I named this `ServerOptions`.  I don't have a strong preference as to which name we go with, however you'll need to change this anyway as there is a facility for sending a custom response back to the client which would need replicating here.

File: Sources/KituraNet/HTTP/HTTPServer.swift [hunk 3]
```diff
@@ -127,22 +129,35 @@ public class HTTPServer: Server {
 
     var quiescingHelper: ServerQuiescingHelper?
 
+    private var ctx: ChannelHandlerContext?
+    
+    /// server configuration
+    public var serverConfig: HTTPServerConfiguration
```
Comment (djones6) [inline Sources/KituraNet/HTTP/HTTPServer.swift]: This needs to be consistent with Kitura-net:
```swift
public var options: ServerOptions
```

File: Sources/KituraNet/HTTP/HTTPServer.swift [hunk 4]
```diff
@@ -127,22 +129,35 @@ public class HTTPServer: Server {
...
      ### Usage Example: ###
      ````swift
-     let server = HTTPServer()
+     let config =HTTPServerConfiguration(requestSize: 1000, coonectionLimit: 100)
+     let server = HTTPServer(serverconfig: config)
      server.listen(on: 8080)
      ````
     */
-    public init() {
+    public init(serverConfig: HTTPServerConfiguration = .default) {
```
Comment (djones6) [inline Sources/KituraNet/HTTP/HTTPServer.swift]: Kitura-net's initializer does not take a parameter.  Instead, `HTTPServer.options` has an initial value of the default options, and is set by the user after the server is created, if desired.
(See usage in https://github.com/IBM-Swift/Kitura/pull/1481)

File: Sources/KituraNet/HTTP/HTTPRequestHandler.swift [hunk 5]
```diff
@@ -76,12 +77,42 @@ internal class HTTPRequestHandler: ChannelInboundHandler, RemovableChannelHandle
...
-
         switch request {
         case .head(let header):
+            serverRequest = HTTPServerRequest(channel: context.channel, requestHead: header, enableSSL: enableSSLVerification)
+            if let requestSizeLimit = server.options.requestSizeLimit,
+                let contentLength = header.headers["Content-Length"].first,
+                let contentLengthValue = Int(contentLength) {
+                if contentLengthValue > requestSizeLimit {
+                    do {
+                        if let (httpStatus, response) = server.options.requestSizeResponseGenerator(requestSizeLimit, "") {
```
Comment (djones6) [inline Sources/KituraNet/HTTP/HTTPRequestHandler.swift]: You are passing an empty string here for the client's identity.  This is used by the default response generator to log a message indicating which client is being rejected.  For example, Kitura-net uses the remote IP address + port of the client.

File: Sources/KituraNet/HTTP/HTTPRequestHandler.swift [hunk 6]
```diff
@@ -152,4 +199,20 @@ internal class HTTPRequestHandler: ChannelInboundHandler, RemovableChannelHandle
     func updateKeepAliveState() {
         keepAliveState.decrement()
     }
+
+    func channelInactive(context: ChannelHandlerContext, httpServer: HTTPServer) {
+        httpServer.connectionCount.sub(1)
+    }
+    
+    func getHeaderSize(of header: HTTPRequestHead) -> Int {
```
Comment (djones6) [inline Sources/KituraNet/HTTP/HTTPRequestHandler.swift]: Could this function be removed now?

File: Tests/KituraNetTests/LargePayloadTests.swift [hunk 7]
```diff
@@ -80,13 +80,15 @@ class LargePayloadTests: KituraNetTest {
 
         func handle(request: ServerRequest, response: ServerResponse) {
             if  request.method.uppercased() == "GET" {
+                print("hello world")
```
Comment (djones6) [inline Tests/KituraNetTests/LargePayloadTests.swift]: These print statements probably want to be removed?

File: Tests/KituraNetTests/KituraNetWebSocketUpgrade.swift [hunk 8]
```diff
@@ -31,7 +31,7 @@ class KituraNetWebSocketUpgradeTest: KituraNetTest {
         ]
     }
     
-    var httpHandler: HTTPResponseHandler?
+    var httpHandler: ResponseHandler?
```
Comment (djones6) [inline Tests/KituraNetTests/KituraNetWebSocketUpgrade.swift]: This rename doesn't seem to belong to the same set of changes

File: Tests/KituraNetTests/ConnectionLimitTests.swift [hunk 9]
```diff
@@ -0,0 +1,93 @@
...
+        do {
+            try channel = bootstrap.connect(host: "localhost", port: self.port).wait()
+            let request = HTTPRequestHead(version: HTTPVersion(major: 1, minor: 1), method: .GET, uri: "/")
+            self.sendRequest(request: request, on: channel)
+        } catch _ {
+            XCTFail("Connection is not established.")
+        }
+    }
+
+func testConnectionLimit() {
```
Comment (djones6) [inline Tests/KituraNetTests/ConnectionLimitTests.swift]: Can you fix the indentation for the rest of this file? It's difficult to tell whether (for example) this function and the classes defined below are within the scope of `ConnectionLimitTests` or not.

File: Sources/KituraNet/HTTP/RequestsizeHandler.swift [hunk 10]
```diff
@@ -0,0 +1 @@
+
```
Comment (djones6) [inline Sources/KituraNet/HTTP/RequestsizeHandler.swift]: This appears to be a new file with only whitespace in it, can it be removed?

File: Sources/KituraNet/HTTP/HTTPServer.swift [hunk 11]
```diff
@@ -309,15 +324,16 @@ public class HTTPServer: Server {
             }
             .childChannelInitializer { channel in
                 let httpHandler = HTTPRequestHandler(for: self)
-                let config: NIOHTTPServerUpgradeConfiguration = (upgraders: upgraders, completionHandler: { _ in
+                let config: HTTPUpgradeConfiguration = (upgraders: upgraders, completionHandler: { ctx in
+                    self.ctx = ctx
```
Comment (djones6) [inline Sources/KituraNet/HTTP/HTTPServer.swift]: Does this new property `ctx` ever get read? I can't see reference to it elsewhere in this PR

File: Sources/KituraNet/HTTP/HTTPServer.swift [hunk 12]
```diff
@@ -309,15 +324,16 @@ public class HTTPServer: Server {
...
                     _ = channel.pipeline.removeHandler(httpHandler)
                 })
                 return channel.pipeline.configureHTTPServerPipeline(withServerUpgrade: config, withErrorHandling: true).flatMap {
                     if let nioSSLServerHandler = self.createNIOSSLServerHandler() {
                         _ = channel.pipeline.addHandler(nioSSLServerHandler, position: .first)
                     }
                     return channel.pipeline.addHandler(httpHandler)
-                }
+                    }
```
Comment (djones6) [inline Sources/KituraNet/HTTP/HTTPServer.swift]: whitespace error

**Comment by RudraniW** [comment]
@djones6 Thanks for your feedback.
I have made the changes you asked.

File: Sources/KituraNet/HTTP/HTTPServer.swift [hunk 13]
```diff
@@ -127,22 +129,35 @@ public class HTTPServer: Server {
 
     var quiescingHelper: ServerQuiescingHelper?
 
+    private var ctx: ChannelHandlerContext?
```
Comment (djones6) [inline Sources/KituraNet/HTTP/HTTPServer.swift]: Please remove this unused var

File: Tests/KituraNetTests/LargePayloadTests.swift [hunk 14]
```diff
@@ -87,6 +87,7 @@ class LargePayloadTests: KituraNetTest {
         }
 
         func handleGet(request: ServerRequest, response: ServerResponse) {
+            print("handle Get")
```
Comment (djones6) [inline Tests/KituraNetTests/LargePayloadTests.swift]: Did you intend to remove this one too?  The only other change in this file is a whitespace change, so I'd suggest just reverting this file.

File: Sources/KituraNet/HTTP/HTTPServer.swift [hunk 15]
```diff
@@ -127,22 +129,35 @@ public class HTTPServer: Server {
...
+    
+    /// server configuration
+    public var options: ServerOptions = ServerOptions()
+
+    //counter for no of connections
+    var connectionCount = Atomic(value: 0)
+
+    // The data to be written as a part of the response.
+    //private var buffer: ByteBuffer
```
Comment (djones6) [inline Sources/KituraNet/HTTP/HTTPServer.swift]: Remove this commented out code

**Pull request merged**